File: dagger2/src/main/java/pgreze/dagger2/app/Dagger2App.java
package pgreze.dagger2.app;

import android.app.Application;
import android.content.Context;

import javax.inject.Inject;

import pgreze.dagger2.analytics.AnalyticsManager;

public class Dagger2App extends Application {

    public static Dagger2App getInstance(Context context) {
        return (Dagger2App) context.getApplicationContext();
    }

    private AppComponent component;

    @Inject AnalyticsManager analyticsManager;

    @Override
    public void onCreate() {
        super.onCreate();
        setupGraph();
        analyticsManager.event("app_opened");
    }

    private void setupGraph() {
        component = DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .build();
        component.inject(this);
    }

    public AppComponent component() {
        return component;
    }
}

File: dagger/android/src/main/kotlin/pgreze/dagger2/app/Dagger2App.kt
package pgreze.dagger2.app

import android.app.Application
import android.content.Context

import javax.inject.Inject

import pgreze.dagger2.analytics.AnalyticsManager
import pgreze.dagger2.di.HasComponent

class Dagger2App : Application(), HasComponent<AppComponent> {

    @Inject
    internal lateinit var analyticsManager : AnalyticsManager

    override val component: AppComponent by lazy {
        DaggerAppComponent.builder()
                .appModule(AppModule(this))
                .build()
    }

    override fun onCreate() {
        super.onCreate()
        component.inject(this)
        analyticsManager.event("App Opened")
    }

    companion object {

        fun getInstance(context: Context): Dagger2App {
            return context.applicationContext as Dagger2App
        }
    }
}
